cmd/mel-replay/receipt_fetcher.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2025-2026, Offchain Labs, Inc.
+// For license information, see https://github.com/OffchainLabs/nitro/blob/master/LICENSE.md
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rlp"
+
+	"github.com/offchainlabs/nitro/arbutil"
+)
+
+type receiptFetcherForBlock struct {
+	header           *types.Header
+	preimageResolver preimageResolver
+}
+
+// ReceiptForTransactionIndex fetches a receipt for a specific transaction index by walking
+// the receipt trie of the block header. It uses the preimage resolver to fetch the preimages
+// of the trie nodes as needed.
+func (rf *receiptFetcherForBlock) ReceiptForTransactionIndex(
+	ctx context.Context,
+	txIndex uint,
+) (*types.Receipt, error) {
+	return fetchReceiptFromBlock(rf.header.ReceiptHash, txIndex, rf.preimageResolver)
+}
+
+// Fetches a specific receipt index from a block's receipt trie by navigating its
+// Merkle Patricia Trie structure. It uses the preimage resolver to fetch preimages
```

Comment:
This is not a comment that requires any action. I'm just a bit surprised that we don't have a golang data structure lying around that provides a more elegant API to navigating a MPT. I wonder how many places in our codebase we do a similar search and have to do all the special handling of navigating the structure.

Reply:
The only utility we have is a trie.Iterator(), which will give you all the nodes in the tree, but that is not sufficient when we want a specific path along the tree. There aren't any other places in the codebase where we really need to do this, and the existing replay binary has no need either